### PR TITLE
Feature/allergies stu3 review

### PR DESCRIPTION
**Pull request by au2**
#### What does this PR do?

1) Changes “Allergy Concern Act StatusCode” to  “AllergyIntolerance.verificationStatus” mapping and adds unit test.
2) Adds “AllergyIntolerance.ClinicalStatus” implementation by mapping “ProblemStatus” and adds a unit test.
3) Adds “AllergyIntolerance.lastOccurance” implementation by using the latest of reaction onset datetimes and adds a unit test.
4) Changes “AlergyIntolerance.onset” to “AllergyIIntolerance.onsetDatetime”, changes source to “Allergy Concern Observation EffectiveTime” based on descriptions and adds a unit test.
5) Adds “AllergyIntolerance.assertedDate” implementation, uses “Allergy Concern Act EffectiveTime” as source and adds unit test.
6) Adds unit test or “AllergyIntolerance.code”.
7) Changes “AllergyIntolerance.category” mapped values and adds a unit test.
8) Comments a number of tickets as unneeded.

#### Related JIRA tickets:

UPMCFHIR-28
UPMCFHIR-29
UPMCFHIR-30
UPMCFHIR-31
UPMCFHIR-32
UPMCFHIR-33
UPMCFHIR-35
UPMCFHIR-38
UPMCFHIR-40

#### How should this be manually tested?

N/A

#### Background/Context:

AllergyIntolerance resource STU3 first review.

#### New JIRA tickets for clinical review required?

Possibly for UPMCFHIR-29 verificationStatus and UPMCFHIR-30 category.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -44,7 +44,7 @@
         <dependency>
             <groupId>org.openehealth.ipf.modules</groupId>
             <artifactId>ipf-modules-cda-mdht</artifactId>
-            <version>3.5.2</version>
```
Comment (mmccall) [inline pom.xml]: was this rolled forward for a reason?

File: pom.xml [hunk 2]
```diff
@@ -96,5 +96,15 @@
             <scope>system</scope>
             <systemPath>${project.basedir}/lib/org.hl7.fhir.validator.jar</systemPath>
         </dependency>
+		<dependency>
```
Comment (mmccall) [inline pom.xml]: If we are doing all of our Jolt work on the repo this feeds into, can we back this out or are you using the JSON utils for anything?

File: src/main/java/tr/com/srdc/cda2fhir/transform/IValueSetsTransformer.java [hunk 3]
```diff
@@ -252,4 +253,12 @@
 	 * @return A value from the FHIR valueset ContactPointSystem.
 	 */
 	ContactPointSystem tTelValue2ContactPointSystem(String cdaTelValue);
+	
+	/**
+	 * Transforms a CDA ProblemStatus code to a AllergyIntoleranceClinicalStatus
+	 * @param code ProblemStatus code
+	 * @return A value from the FHIR AllergyIntoleranceClinicalStatus 
+	 */
+	AllergyIntoleranceClinicalStatus tProblemStatus2AllergyIntoleranceClinicalStatus(String code);
```
Comment (mmccall) [inline src/main/java/tr/com/srdc/cda2fhir/transform/IValueSetsTransformer.java]: does it make sense to generalize this function to be tProblemStatus2ClinicalStatus?  The AllergyIntolerance status codes look like a subset of the problem status codes, which we will need to update anyways once we tackle that resource.

File: pom.xml (hunk 2 above)
Comment (mmccall) [reply pom.xml]: nevermind, looks like you are using jolt, we can address this later.

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    	return allergyResources.get(0);	
+	}
+	
+	static private void verifyAllergyIntoleranceCategory(AllergyProblemAct act, String expected) throws Exception {
+		Bundle bundle = rt.tAllergyProblemAct2AllergyIntolerance(act);
+		AllergyIntolerance allergyIntolerance = findOneResource(bundle);
+
+    	Enumeration<AllergyIntolerance.AllergyIntoleranceCategory> category = allergyIntolerance.getCategory().get(0);
+    	String actual = category == null ? null : category.asStringValue();
+		Assert.assertEquals(expected, actual);
```
Comment (mmccall) [inline src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: do you mind putting in textual descriptions of the asserts similar to how I did in my PR? Helps to just define the assertions themselves.

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+		
+		cdaObjFactory = (ConsolFactoryImpl) ConsolFactoryImpl.init();
+		cdaTypeFactory = DatatypesFactoryImpl.init();		
+		cdaFactory = (CDAFactoryImpl) CDAFactoryImpl.init();
+		
+		cdaProblemStatusCodeToName.put("55561003", "Active");
+		cdaProblemStatusCodeToName.put("73425007", "Inactive");
+		cdaProblemStatusCodeToName.put("413322009", "Resolved");
+		
+		cdaAllergyIntolaranceTypeCodeToName.put("419199007", "Allergy to substance (disorder)");
```
Comment (mmccall) [inline src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: you misspelled Intolerance here. Not a big deal, but fix if you want.

File: pom.xml (hunk 1 above)
Comment (au2) [reply pom.xml]: They changed the name of the Act to Allergy Concern Act in the spec but not in the code (still Allergy Problem Act). I thought the latest version (January 2019) would have hat change but did not. I left it since using the latest version is probably better.

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+		Bundle bundle = rt.tAllergyProblemAct2AllergyIntolerance(act);
+		AllergyIntolerance allergyIntolerance = findOneResource(bundle);
+		
+    	AllergyIntoleranceVerificationStatus verificationStatus = allergyIntolerance.getVerificationStatus();
+    	String actual = verificationStatus == null ? null : verificationStatus.toCode();
+		Assert.assertEquals(expected, actual);		
+	}
+	
+	@Test
+	public void testAllergyIntoleranceStatusCode() throws Exception {
```
Comment (mmccall) [inline src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java:373]: would you mind adding a test/assertion here for statusCode NULL. Just want to make sure we aren't defaulting anything to active when nothing arrives on the record. It isn't required per the spec, and doesn't seem to be on any of the test CCDAs that come with the lib.

File: pom.xml (hunk 2 above)
Comment (au2) [reply pom.xml]: Right. I use it for JSON parsing. Gives out a nice Map.

File: src/main/java/tr/com/srdc/cda2fhir/transform/IValueSetsTransformer.java (hunk 3 above)
Comment (au2) [reply src/main/java/tr/com/srdc/cda2fhir/transform/IValueSetsTransformer.java]: I will do the conditions next. Can take a look at that point.

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java (hunk 4 above)
Comment (au2) [reply src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: ok

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java (hunk 5 above)
Comment (au2) [reply src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: ok

**Review by mmccall — COMMENTED** [review]
Overall looks great to me, just a few questions here and there.

File: pom.xml (hunk 1 above)
Comment (mmccall) [reply pom.xml]: ok

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java (hunk 6 above)
Comment (au2) [reply src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java:373]: there was one and added another one after specifically setting status code to zero.

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java (hunk 5 above)
Comment (au2) [reply src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: corrected

File: src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java (hunk 4 above)
Comment (au2) [reply src/test/java/tr/com/srdc/cda2fhir/AllergyConcernActTest.java]: i put messages

**Comment by mmccall** [comment]
lgtm.

**Review by mmccall — APPROVED** [review]
